plugins/optimization-detective/class-od-url-metric-group-collection.php

```diff
@@ -396,34 +397,61 @@ public function get_common_lcp_element(): ?array {
+	 * (the default sample size). Therefore, given the number (n) of visited elements on the page this will only
+	 * end up running n*4*3 times.
+	 *
+	 * @todo Should there be an OD_Element class which has a $url_metric property which then in turn has a $group property. Then this would only need to return array<string, OD_Element[]>.
+	 * @since n.e.x.t
+	 *
+	 * @return array<string, non-empty-array<int, array{OD_URL_Metric_Group, OD_URL_Metric, ElementData}>> Keys are XPaths and values are arrays of tuples consisting of the group, URL metric, and element data.
 	 */
-	public function get_all_element_max_intersection_ratios(): array {
+	public function get_all_denormalized_elements(): array {
 		if ( array_key_exists( __FUNCTION__, $this->result_cache ) ) {
 			return $this->result_cache[ __FUNCTION__ ];
 		}
 
 		$result = ( function () {
-			$element_max_intersection_ratios = array();
-
-			/*
-			 * O(n^3) my! Yes. This is why the result is cached. This being said, the number of groups should be 4 (one
-			 * more than the default number of breakpoints) and the number of URL metrics for each group should be 3
-			 * (the default sample size). Therefore, given the number (n) of visited elements on the page this will only
-			 * end up running n*4*3 times.
-			 */
+			$all_denormalized_elements = array();
 			foreach ( $this->groups as $group ) {
 				foreach ( $group as $url_metric ) {
 					foreach ( $url_metric->get_elements() as $element ) {
-						$element_max_intersection_ratios[ $element['xpath'] ] = array_key_exists( $element['xpath'], $element_max_intersection_ratios )
-							? max( $element_max_intersection_ratios[ $element['xpath'] ], $element['intersectionRatio'] )
-							: $element['intersectionRatio'];
+						$all_denormalized_elements[ $element['xpath'] ][] = array( $group, $url_metric, $element );
 					}
 				}
 			}
-			return $element_max_intersection_ratios;
+			return $all_denormalized_elements;
+		} )();
+
+		$this->result_cache[ __FUNCTION__ ] = $result;
+		return $result;
+	}
+
+	/**
+	 * Gets the max intersection ratios of all elements across all groups and their captured URL metrics.
+	 *
+	 * @return array<string, float> Keys are XPaths and values are the intersection ratios.
+	 */
+	public function get_all_element_max_intersection_ratios(): array {
+		if ( array_key_exists( __FUNCTION__, $this->result_cache ) ) {
+			return $this->result_cache[ __FUNCTION__ ];
+		}
+
+		$result = ( function () {
+			$elements_max_intersection_ratios = array();
+			foreach ( $this->get_all_denormalized_elements() as $xpath => $denormalized_elements ) {
+				$element_intersection_ratios = array();
+				foreach ( $denormalized_elements as list( $group, $url_metric, $element ) ) {
+					$element_intersection_ratios[] = $element['intersectionRatio'];
+				}
+				$elements_max_intersection_ratios[ $xpath ] = count( $element_intersection_ratios ) > 1
+					? (float) max( ...$element_intersection_ratios )
```

Comment:
Interestingly, if there is only one element in the array than this is a fatal error: https://3v4l.org/nQNZh

Too bad it doesn't just return the initial item.

Reply:
why is the spread needed here, can this just be `(float) max( $element_intersection_ratios )`?

Reply:
Oh, good point. It seems this is an alternate signature:

![image](https://github.com/user-attachments/assets/25777242-3736-4568-8d41-1f726a3ec86e)

Reply:
Hah, and then there's no need for the ternary since a single element array does not cause an error: https://3v4l.org/fmIl9

Reply:
Addressed in dca43e9d8ff8cf121ee5c53e1d8e1903840bedc8